### PR TITLE
8275527: Refactor forward pointer access

### PR DESCRIPTION
**Pull request by rkennke**
Accessing the forward pointer is currently a little inconsistent. Some code paths call oopDesc::forwardee() / oopDesc::is_forwarded(), some code paths call forwardee() and check it for ==/!= NULL, some code paths even call markWord::decode_pointer() and markWord::is_marked() instead.

This change attempts to make the situation more consistent. For simple cases it preserves oopDesc::forwardee() / is_forwarded(), some cases need to use the markWord for consistency in concurrent GC, they now use markWord::forwardee() and markWord::is_forwarded(). Also, checking whether or not an object is forwarded is now consistently done using is_forwarded() and not by checking forwardee ==/!= NULL. This also resolves the mess in G1 full GC that changes not-forwarded objects to have a NULL (fake-) pointer. This is not necessary, because we can just as well use the lock bits to determine whether or not the object is forwarded.

Testing:
 - [x] tier
 - [x] tier2
 - [x] hotspot_gc

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8275527](https://bugs.openjdk.java.net/browse/JDK-8275527): Refactor forward pointer access

### Reviewers
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to a0e3a6a5f9673f574c485a6bdd812c0f7ae090c2
 * [Stefan Karlsson](https://openjdk.java.net/census#stefank) (@stefank - **Reviewer**) ⚠️ Review applies to 9c2ed2ff02ce5f7197ea28c41cdb7b119b7332f1

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5955/head:pull/5955` \
`$ git checkout pull/5955`

Update a local copy of the PR: \
`$ git checkout pull/5955` \
`$ git pull https://git.openjdk.java.net/jdk pull/5955/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5955`

View PR using the GUI difftool: \
`$ git pr show -t 5955`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5955.diff">https://git.openjdk.java.net/jdk/pull/5955.diff</a>

</details>

### CODE REVIEW

**Comment by tschatzl** [comment]
@rkennke : could you close the superfluous CRs for this issue? You filed this one 9 times in total... (probably due to network issues yesterday). Thanks.

**Comment by rkennke** [comment]
> Never mind, already did that. I did not check if any of the duplicates contain more information than the one used for this PR though.

Oh dear! Thank you! Yes, when I tried to file the issue, bugs.o.j.n was down and/or very slow. Apparently every attempt actually went through, without actually sending back a confirmation.

**Comment by rkennke** [comment]
> I like how the mark/is_marked/decode_pointer code gets cleaned up. And having an OopForwading class probably makes sense for parts of the code.
> 
> I'm less excited about the way the explicitly stack allocated OopForwarding objects make the code less readable. Take these two examples:
> 
> ```
>    assert(old->is_objArray(), "invariant");
> -  assert(old->is_forwarded(), "invariant");
> +  assert(OopForwarding(old).is_forwarded(), "invariant");
> ```
> 
> ```
>      oop new_obj = obj->is_forwarded() ? obj->forwardee()
>                                         : _young_gen->copy_to_survivor_space(obj);
>       OopForwarding fwd(obj);
>       oop new_obj = fwd.is_forwarded() ? fwd.forwardee()
>                                        : _young_gen->copy_to_survivor_space(obj);
> ```
> 
> I think both snippets were nicer to read in the earlier code. And there's a lot of those kind of changes in this patch.
> 
> Maybe there's a hybrid approach where we could still have oopDesc::is_forwarded, oopDesc::forwardee, etc. but also have OopForwarding objects for those places that really need them? And then minimize the amount of places we use `OopForwarding fwd(obj)` and `OopForwarding(obj)->forwardee()` in the code?

In the first snippet, there is only one usage of the mark-word, i.e. no problem to simplify that. However, in the 2nd snippet, the point is that we can avoid loading the mark-word twice, and if we want to do that, we need an enclosing scope. The alternative would be to do something like:
markWord mark = obj->mark();
oop fwd = mark->is_forwarded() ? mark->forwardee() : something_else();
... which is pretty much the same as what I proposed, except that the markWord and the logic is encapsulated in OopForwarding.

Notice that in some cases we actually already do that - those are the 'messy' cases where we deliberately load the mark-word once, and drag it through the code that needs it (sometimes not only for performance but also for correctness in the face of concurrency) - and the use is_marked() and decode_pointer() on it.

Given that I have not been able to show performance benefits, I could revert cases like you mentioned to their original form, and consolidate the implementation into markWord, unless we prefer the most efficient solution everywhere. ?

**Comment by stefank** [comment]
I don't think we should make the code more complicated by using the micro optimized version unless we can measure the benefits. IMHO, it is harder to read and it makes it more difficult to know if the read-once of was done for correctness purposes or not. I understand that that other reviewers might disagree with my view, and if that's the case, then I'll back off.

Some concrete feedback on the current patch:

I don't like that we store the oop in the OopForwarding. That oop gets snuck into functions via the fwd variable. That makes it less obvious that the functions depend on the old oop. Code like the following becomes more obscure, when you don't immediately see that it operates on the original (old) object:
```
  const oop forward_ptr = fwd.forward_to_atomic(obj, memory_order_relaxed);
  if (forward_ptr == NULL) {
```

I wonder if it wouldn't make sense to make forward_to_atomic static, just like the forward_to is static? That way we wouldn't have to store _obj in OopForwarding, and I think the code would be more decoupled and easier to read.

File: src/hotspot/share/oops/markWord.hpp [hunk 1]
```diff
@@ -243,7 +243,13 @@ class markWord {
...
   // Recover address of oop from encoded form used in mark
-  inline void* decode_pointer() { return (void*)clear_lock_bits().value(); }
+  inline void* decode_pointer() const { return (void*)clear_lock_bits().value(); }
+
+  inline bool is_forwarded() const { return is_marked(); }
+  inline oop forwardee() const {
+    assert(is_forwarded(), "only decode when actually forwarded");
+    return cast_to_oop(decode_pointer());
+  }
 };
```
Comment (stefank) [inline src/hotspot/share/oops/markWord.hpp:253]: This brings the forwarded/forwardee terminology into the markWord. The markWord was previously decoupled from those to concepts. I would personally let those function names stay in oopDesc and not leak down into the markWord. If you do want to keep it here, could you update the comments at the top that describes the bits?
```
//    [ptr             | 11]  marked             used to mark an object
```

**Review by stefank — APPROVED** [review]
Thanks for doing this change. This looks good to me. I've added a comment below that I think would be nice to get resolved somehow, though I don't need to re-review if you update with any of the suggestions.

File: src/hotspot/share/oops/markWord.hpp (hunk 1 above)
Comment (rkennke) [reply src/hotspot/share/oops/markWord.hpp]: Yeah, I am not quite sure about this. We have a couple of places where we need to use the markWord direcly, and they read m.is_marked() (when it really means is_forwarded, even though it's the same in the implementation), and then goes on to cast_to_oop(m.decode_pointer()) which reads more ugly than simply m.forwardee() which also comes with an assert and the cast.

I reverted the markWord change and related call-sites now. Maybe this warrants more thinking/discussion.

**Pull request closed without merging**